### PR TITLE
fix: fixed active app indicator being shown in full screen spaces and fixed indicator not being aligned right below of active icon when dock size changes due to app being closed

### LINKED ISSUE #899: [FR] hide the active app indicator when cursor is over the dock (genie effect messes with it)
## What problem does this solve?
active app indicator is cool and useful, but if the dock genie effect is on it is inaccurate when the mouse is hovering over dock icons and zooming them

## Proposed solution
hide the active app indicator when the cursor is interacting with the dock

## Additional context (optional)

- [ x] I have reviewed existing issues to ensure this is not a duplicate

### PR DESCRIPTION
**Pull request by maddada**
## Describe your changes

Improved the following aspects related to the active app indicator:

- Indicator should play nicer with minimize effects like genie.
- It appeared in Full Screen spaces.
- It was often shifted when the size of the dock changes due to an app icon being shown or removed due to an app being started or quit.

## Related issue number(s) and link(s)

Closes #899 which had the 3 issues listed above in first post and in 2 comments

2026-01-18_Vivaldi Snapshot_03-21-45

2026-01-18_Vivaldi Snapshot_03-22-00

## Checklist before requesting a review
- [x] I have performed a self-review of my code and understand every line
- [x] If this change affects core functionality, I have added a description highlighting the changes
- [x] I have followed conventional commit guidelines
- [x] I have tested this PR on my own machine

## AI Assistance Disclosure

**Did you use AI tools (ChatGPT, Claude, Copilot, Cursor, etc.) for this PR?**
- [ ] No AI tools were used
- [x] Yes - describe below how AI was used and confirm you reviewed/tested the output

I used Claude Code Opus 4.5 and checked every line and discussed the performance implications and tested the change thoroughly and handled all edge cases. Also asked it to add print statements for all of these event handlers and checked that all of them are not being called excessively or unnecessarily. I'm mostly a full stack web app engineer so forgive me for any lapses in knowledge. Thank you!

## Core Functionality Changes

This fix doesn't modify core functionality just fixes few issues with active app indicator.

### CODE REVIEW

**Comment by maddada** [comment]
### This is just extra info from Claude regarding the events we're using now to position the active app indicator:

### Active App Indicator - Event Subscriptions

This document lists all events/observers used by `ActiveAppIndicatorCoordinator` for the active app indicator feature.

### Summary

| Observer | Event | Status | Frequency | Purpose |
|----------|-------|--------|-----------|---------|
| workspaceObserver | didActivateApplicationNotification | EXISTING | On app switch | Move indicator to active app |
| positionSettingsObserver | Defaults.observe | EXISTING | On settings change | Update indicator appearance |
| screenParametersObserver | didChangeScreenParametersNotification | EXISTING | On display/dock changes | Handle dock resize |
| spaceChangeObserver | activeSpaceDidChangeNotification | NEW | On space/fullscreen change | Hide indicator in fullscreen |
| dockLayoutObserver | AXObserver on Dock list | NEW | On dock icon add/remove | Handle dock layout shifts |

---

### Detailed Breakdown

### 1. workspaceObserver (EXISTING)
- **Event:** `NSWorkspace.didActivateApplicationNotification`
- **Source:** NSWorkspace notification center
- **Fires when:** User switches to a different application
- **Frequency:** Only when active app changes (user-initiated)
- **Performance cost:** Very low - only fires on deliberate app switches
- **Action:** Updates indicator position immediately + schedules delayed update

### 2. positionSettingsObserver (EXISTING)
- **Event:** Defaults.observe on indicator settings keys
- **Source:** Defaults library (UserDefaults wrapper)
- **Fires when:** User changes indicator settings (size, offset, color, etc.)
- **Frequency:** Only when user modifies settings in preferences
- **Performance cost:** Negligible - user-initiated, rare
- **Action:** Repositions indicator with new settings

### 3. screenParametersObserver (EXISTING)
- **Event:** `NSApplication.didChangeScreenParametersNotification`
- **Source:** NotificationCenter
- **Fires when:** Display configuration changes (resolution, arrangement, dock position)
- **Frequency:** Rare - only on display/dock configuration changes
- **Performance cost:** Very low
- **Action:** Checks if dock position/size changed, schedules update if needed

### 4. spaceChangeObserver (NEW - for fullscreen fix)
- **Event:** `NSWorkspace.activeSpaceDidChangeNotification`
- **Source:** NSWorkspace notification center
- **Fires when:** User switches Mission Control spaces or enters/exits fullscreen
- **Frequency:** Only on space changes (user-initiated)
- **Performance cost:** Very low - user-initiated action
- **Action:** Checks if in fullscreen, hides indicator if dock is hidden

### 5. dockLayoutObserver (NEW - for dock shift fix)
- **Event:** `kAXUIElementDestroyedNotification` and `kAXCreatedNotification` on Dock's list element
- **Source:** AXObserver on com.apple.dock process
- **Fires when:** Dock icons are added or removed (app opens window, app quits, etc.)
- **Frequency:** When apps with dock icons launch/quit or apps open/close their first/last window
- **Performance cost:** Low - only fires on actual dock content changes
- **Action:** Schedules delayed indicator update to account for dock animation

---

### Performance Notes

### Events NOT used (alternatives considered but rejected):
- **Polling timer:** Would continuously check dock state - rejected for battery/CPU impact
- **kAXLayoutChangedNotification:** Fires too frequently on dock hover animations
- **Window observer on all apps:** Would require observing every app's windows - too broad

### Why these events are efficient:
1. All observers are event-driven (no polling)
2. Each event fires only when a relevant change occurs
3. Most events are user-initiated (app switch, space change, settings change)
4. The dock layout observer only fires when dock content actually changes, not on hover/animation

### Delayed update mechanism:
- When events fire that might cause dock animation (app switch, dock layout change), we:
  1. Update indicator immediately (for instant feedback)
  2. Schedule a second update after 0.6s (to catch dock animation completion)
- The delayed timer is debounced - multiple rapid events only trigger one delayed update

---

### Testing Notes

**Tested on:** January 2026

**Result:** All events verified to fire only when expected - no excessive event calls observed.

Test scenarios performed:
- Normal app switching (clicking dock icons, Cmd+Tab)
- Hovering over dock icons
- Opening/closing app windows (BTT settings edge case)
- Entering/exiting fullscreen
- Idle state (no spurious events)

All 6 observers behave correctly and only fire on relevant state changes.

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift [hunk 1]
```diff
@@ -77,6 +85,109 @@ final class ActiveAppIndicatorCoordinator {
...
+        } else {
+            // Dock is hidden - hide the indicator
+            indicatorWindow?.orderOut(self)
+        }
+    }
+
+    /// Determines if the dock is currently visible by checking:
+    /// 1. If the frontmost window is fullscreen (dock hidden in fullscreen spaces)
+    /// 2. If the dock size is effectively 0 (auto-hide triggered)
+    private func isDockVisible() -> Bool {
```
Comment (ejbills) [inline DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: I feel like these dock utilities belong in DockObserver. Also, I think you could just call getDockSize from DockUtils and if it returns a value 0 then you know it's not visible.

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift [hunk 2]
```diff
@@ -77,6 +85,109 @@ final class ActiveAppIndicatorCoordinator {
...
+        // Check if frontmost app's focused window is fullscreen
+        if let app = NSWorkspace.shared.frontmostApplication {
+            let appElement = AXUIElementCreateApplication(app.processIdentifier)
+            if let focusedWindow = try? appElement.focusedWindow(),
+               let isFullscreen = try? focusedWindow.isFullscreen(),
```
Comment (ejbills) [inline DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: WindowUtil - isAppInFullscreen is more applicable here

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift [hunk 3]
```diff
@@ -86,12 +197,23 @@ final class ActiveAppIndicatorCoordinator {
...
+        }
+        dockLayoutObserver = nil
+        observedDockList = nil
+        delayedUpdateTimer?.invalidate()
         positionSettingsObserver?.invalidate()
         hideIndicator()
     }
 
     private func handleScreenParametersChanged() {
+        print("[Indicator:screenParametersObserver] Screen parameters changed")
```
Comment (ejbills) [inline DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: Left over?

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift [hunk 4]
```diff
@@ -77,6 +85,109 @@ final class ActiveAppIndicatorCoordinator {
         ) { [weak self] _ in
             self?.handleScreenParametersChanged()
         }
+
+        // Observe space changes (detects fullscreen transitions)
+        spaceChangeObserver = NSWorkspace.shared.notificationCenter.addObserver(
```
[the remark this answers is not in the document]
Comment (ejbills) [reply DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: Honestly, we should just make a clear decision that the WindowManipulationObs is the service that orchestrates these other services. What do you think.

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift (hunk 3 above)
Comment (maddada) [reply DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: Sorry, took it out.

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift (hunk 2 above)
Comment (maddada) [reply DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: Fixed

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift [hunk 5]
```diff
@@ -77,6 +84,80 @@ final class ActiveAppIndicatorCoordinator {
...
+
+    private func setupDockLayoutObserver() {
+        guard let dockApp = NSRunningApplication.runningApplications(withBundleIdentifier: "com.apple.dock").first else {
+            return
+        }
+
+        let dockPID = dockApp.processIdentifier
+        let dockElement = AXUIElementCreateApplication(dockPID)
+
+        // Find the dock's list element
```
Comment (ejbills) [inline DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: ai likes to leave these overly word-y comments, just run a 'remove useless comments' prompt or something. or, for this proj, add a claude rule that comments should only be inserted in areas where the code is not readable or something like that

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift [hunk 6]
```diff
@@ -158,25 +259,41 @@ final class ActiveAppIndicatorCoordinator {
     // MARK: - Active App Handling
 
     private func handleActiveAppChanged(_ app: NSRunningApplication) {
+        print("[Indicator:workspaceObserver] Active app changed: \(app.localizedName ?? "unknown")")
```
Comment (ejbills) [inline DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: more prints

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift (hunk 5 above)
Comment (maddada) [reply DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: Hey, I don't remove them + prints to make the code reviews and further changes by other AIs easier (they're easy to remove later). Will make sure to not include them for this repos PRs in the future 👍🏻

File: DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift [hunk 7]
```diff
@@ -77,6 +84,80 @@ final class ActiveAppIndicatorCoordinator {
...
+        AXObserverAddNotification(observer, dockList, kAXCreatedNotification as CFString, refcon)
+
+        CFRunLoopAddSource(CFRunLoopGetMain(), AXObserverGetRunLoopSource(observer), .defaultMode)
+
+        dockLayoutObserver = observer
+        observedDockList = dockList
+    }
+
+    private func handleDockLayoutChanged() {
+        print("[Indicator:dockLayoutObserver] Dock layout changed (icon added/removed)")
```
[the remark this answers is not in the document]
Comment (maddada) [reply DockDoor/Utilities/ActiveAppIndicator/ActiveAppIndicatorCoordinator.swift]: Removed!

**Comment by maddada** [comment]
> some comments left over and some minor coding guidelines requested

All print statements and print statements removed. Also added guards as requested.

**Comment by ejbills** [comment]
@maddada you didnt push anything new

**Comment by maddada** [comment]
Sorry just pushed
